File: ICPC/2021-2022 ICPC Northwestern European Regional Programming Contest (NWERC 2021)/H_Heating_Up.cpp
#include <bits/stdc++.h>
using namespace std;
const int MAXN = 1000005;

long long a[MAXN], n;
pair<long long, int> v[MAXN];

bool check(long long mid)
{
    vector<bool> vis(MAXN, 0);
    for (int j = 1; j <= n; j++)
    {
        if (mid < v[j].first)
            return false;
        else if (!vis[v[j].second])
        {
            int i = v[j].second;
            vis[v[j].second] = true;
            int l = i - 1, r = i + 1;
            if (l <= 0)
                l += n;
            if (r > n)
                r %= n;
            long long count = mid + a[i];
            bool ch = true;
            while (l != r)
            {
                bool kt_temp = false;
                if (count >= a[l])
                {
                    vis[l] = true;
                    count += a[l];
                    kt_temp = true;
                    l--;
                    if (l <= 0)
                        l = n;
                }
                else if (count >= a[r])
                {
                    count += a[r];
                    vis[r] = true;
                    kt_temp = true;
                    r++;
                    if (r > n)
                        r %= n;
                }
                if (!kt_temp)
                {
                    ch = false;
                    break;
                }
            }
            if (ch && count >= a[l])
                return true;
        }
    }
    return false;
}

int main()
{
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);
    cout.tie(NULL);

    cin >> n;
    for (int i = 1; i <= n; i++)
    {
        cin >> a[i];
        v[i] = {a[i], i};
    }
    sort(v + 1, v + n + 1);

    long long l = 0, r = 1e13;
    while (l < r)
    {
        long long mid = (l + r) / 2;
        if (check(mid))
            r = mid;
        else
            l = mid + 1;
    }
    long long mid = (l + r) / 2;
    if (mid < 0)
        mid = 0;
    while (!check(mid))
        mid++;
    while (mid - 1 >= 0 && check(mid - 1))
        mid--;
    cout << mid;

    return 0;
}

File: ICPC/2021-2022 ICPC Northwestern European Regional Programming Contest (NWERC 2021)/H_Heating_Up.py

from collections import defaultdict
n = int(input())
a = list(map(int, input().split()))

attemped = set()
l = 0
r = max(a)


def checkPossible(initialValue, initialIndex):
    left = initialIndex
    right = (initialIndex+1) % n
    for i in range(0, n):
        if a[left] <= initialValue:
            initialValue += a[left]
            attemped.add(left)
            left = (left-1+n) % n
        elif a[right] <= initialValue:
            initialValue += a[right]
            attemped.add(right)
            right = (right+1+n) % n
        else:
            return False
    return True


def possible(initialValue):
    attemped.clear()
    for i in range(n):
        if i not in attemped and checkPossible(initialValue, i):
            return True
    return False


while l <= r:
    m = (l+r)//2
    if possible(m):
        r = m-1
    else:
        l = m+1

print(l)


"""
n = int(input())
a = list(map(int, input().split()))
v = [(a[i], i+1) for i in range(n)]
v.sort()

def check(mid):
    vis = [False] * (n+1)
    for j in range(1, n+1):
        if mid < v[j-1][0]:
            return False
        elif not vis[v[j-1][1]]:
            i = v[j-1][1]
            vis[v[j-1][1]] = True
            l = i - 1
            r = i + 1
            if l <= 0:
                l += n
            if r > n:
                r %= n
            count = mid + a[i-1]
            ch = True
            while l != r:
                kt_temp = False
                if count >= a[l-1]:
                    vis[l] = True
                    count += a[l-1]
                    kt_temp = True
                    l -= 1
                    if l <= 0:
                        l = n
                elif count >= a[r-1]:
                    count += a[r-1]
                    vis[r] = True
                    kt_temp = True
                    r += 1
                    if r > n:
                        r %= n
                if not kt_temp:
                    ch = False
                    break
            if ch and count >= a[l-1]:
                return True
    return False

l = 0
r = 10**13
while l < r:
    mid = (l + r) // 2
    if check(mid):
        r = mid
    else:
        l = mid + 1

mid = (l + r) // 2
if mid < 0:
    mid = 0
while not check(mid):
    mid += 1
while mid - 1 >= 0 and check(mid - 1):
    mid -= 1

print(mid)

"""
